pkg/scheduler/framework/plugins/interpodaffinity/plugin.go

```diff
@@ -221,11 +221,35 @@ func (pl *InterPodAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod
 		return framework.Queue, err
 	}
 
+	// When queuing this Pod:
+	// - 1. A new node is added with the pod affinity topologyKey, the pod may become schedulable.
+	// - 2. The original node does not have the pod affinity topologyKey but the modified node does, the pod may become schedulable.
+	// - 3. Both the original and modified nodes have the pod affinity topologyKey and they differ, the pod may become schedulable.
 	for _, term := range terms {
-		if _, ok := modifiedNode.Labels[term.TopologyKey]; ok {
-			logger.V(5).Info("a node with matched pod affinity topologyKey was added/updated and it may make pod schedulable",
+		if originalNode == nil {
+			if _, ok := modifiedNode.Labels[term.TopologyKey]; ok {
+				// Case 1: A new node is added with the pod affinity topologyKey.
+				logger.V(5).Info("A node with a matched pod affinity topologyKey was added and it may make the pod schedulable",
+					"pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+				return framework.Queue, nil
+			}
+			continue
+		}
+		originalTopologyValue, originalHasKey := originalNode.Labels[term.TopologyKey]
+		modifiedTopologyValue, modifiedHasKey := modifiedNode.Labels[term.TopologyKey]
+
+		if !originalHasKey && modifiedHasKey {
+			// Case 2: Original node does not have the pod affinity topologyKey, but the modified node does.
+			logger.V(5).Info("A node got updated to have the topology key of pod affinity, which may make the pod schedulable",
 				"pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
-			return framework.Queue, err
+			return framework.Queue, nil
+		}
+
+		if originalHasKey && modifiedHasKey && (originalTopologyValue != modifiedTopologyValue) {
```

Comment:
https://github.com/kubernetes/kubernetes/pull/127470/files#r1806395330
So it seems that we won't encounter it here? @sanposhiho

Reply:
We will; the real world scenario would be like:

1. Pod-a has PodAffinity for zone topology.
2. There are two zone; node1 and node2 are in zone1, while node3 and node4 are in zone2.
3. node1 has a scheduled Pod matching Pod-a's podaffinity, meaning, the Pod-a can go to zone1 (node1 and node2). node3 and node4 will be rejected by PodAffinity plugin.
4. Let's say node1 and node2 are rejected by other plugins, like the resource fit plugin. 
5. As a result, node1/node2 are unschedulable due to the resource fit plugin, while node3/node4 are unschedulable due to PodAffinity plugin. The Pod-a will be back to the unsched queue.
6. Here, let's say node2's zone label is changed and now node2 is moved from zone1 to zone2. It makes the pod schedulable to zone2, instead of zone1. We should move the Pod-a to the activeQ/backoffQ because now node3 or node4 may be able to accept Pod-a.